rust/common-ifc/src/policy.rs

```diff
@@ -0,0 +1,166 @@
+    Result,
+};
+use std::collections::BTreeMap;
+
+type PolicyMapInner<T> = BTreeMap<T, Context>;
+
+/// Map of [Confidentiality] or [Integrity]
+/// labels to the minimum allowed [Context].
+///
+/// Map is validated upon construction.
+struct PolicyMap<T: Lattice + 'static>(PolicyMapInner<T>);
+
+impl<T> std::ops::Deref for PolicyMap<T>
+where
+    T: Lattice + 'static,
+{
+    type Target = PolicyMapInner<T>;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<T> TryFrom<PolicyMapInner<T>> for PolicyMap<T>
+where
+    T: Lattice + 'static,
+{
+    type Error = PolicyError;
+
+    /// Constructs a new policy map, validating that
+    /// all labels have defined requirements.
+    fn try_from(map: PolicyMapInner<T>) -> Result<Self> {
+        for label in T::iter() {
+            if !map.contains_key(label) {
+                return Err(PolicyError::InvalidPolicy(format!(
+                    "No requirements defined for {label}"
+                )));
+            }
+        }
+        Ok(Self(map))
+    }
+}
+
+/// Represents an invoker's data flow requirements
+/// to verify an execution graph with regard to the provided
+/// data input.
+///
+/// Internally, contains maps of [Confidentiality]
+/// and [Integrity] levels to [Context] requirements.
+pub struct Policy {
+    /// Map of confidentiality principals to the minimum
+    /// required [Context] components.
+    confidentiality_map: PolicyMap<Confidentiality>,
+    /// Map of integrity principals to the minimum
+    /// required [Context] components.
+    integrity_map: PolicyMap<Integrity>,
+}
+
+impl Policy {
+    /// Create a new [Policy], given [PolicyMap] for integrity
+    /// and confidentiality.
+    pub fn new<C, I>(confidentiality_map: C, integrity_map: I) -> Result<Self>
+    where
+        C: Into<PolicyMapInner<Confidentiality>>,
+        I: Into<PolicyMapInner<Integrity>>,
+    {
+        let confidentiality_map = confidentiality_map.into().try_into()?;
+        let integrity_map = integrity_map.into().try_into()?;
+        Ok(Self {
+            confidentiality_map,
+            integrity_map,
+        })
+    }
+
+    /// Create a [Policy] from defaults.
+    ///
+    /// Explicitly not using [Default] trait so that
+    /// defaults go through the same validation.
+    pub fn with_defaults() -> Result<Self> {
+        use Confidentiality::*;
+        use Integrity::*;
+        use ModuleEnvironment::*;
+
+        let confidentiality_map = [(Public, (Server,).into()), (Private, (Server,).into())];
+        let integrity_map = [
+            (LowIntegrity, (Server,).into()),
+            (HighIntegrity, (Server,).into()),
+        ];
+
+        Self::new(confidentiality_map, integrity_map)
+    }
+
+    /// Validate input against this policy, given a [Context].
+    pub fn validate<'a, T, I>(&'a self, input: I, ctx: &Context) -> Result<()>
+    where
+        T: 'static,
+        I: Iterator<Item = (&'a String, &'a Data<T>)>,
```

Comment:
It may be more ergonomic to accept [`IntoIterator<Item = ...>`](https://doc.rust-lang.org/std/iter/trait.IntoIterator.html) here

Reply:
Indeed this is what I want